sklearn/cluster/optics_.py

```diff
@@ -0,0 +1,574 @@
+
+        Parameters
+        ----------
+        X : array, shape (n_samples, n_features)
+            The data.
+
+        Returns
+        -------
+        self : instance of OPTICS
+            The instance.
+        """
+        X = check_array(X)
+
+        n_samples = len(X)
+
+        # Check for valid n_samples relative to min_samples
+        if self.min_samples > n_samples:
+            raise ValueError("Number of training samples (%d) must be greater "
+                             "than min_samples (%d) used for clustering." %
+                             (n_samples, self.min_samples))
+
+        self.tree_ = _SetOfObjects(X, self.metric)
+        _prep_optics(self.tree_, self.min_samples)
+        _build_optics(self.tree_, self.eps * 5.0)
+        self.extract_auto()
+        self.core_sample_indices_ = np.arange(n_samples)[self._is_core]
+        self.n_clusters_ = max(self._cluster_id)
+        return self
+
+    @property
+    def reachability_(self):
+        return self.tree_.reachability_[:]
+
+    @property
+    def core_dists_(self):
+        return self.tree_.core_dists_[:]
+
+    @property
+    def _cluster_id(self):
+        return self.tree_._cluster_id[:]
+
+    @property
+    def _is_core(self):
+        return self.tree_._is_core[:]
+
+    @property
+    def _index(self):
+        return self.tree_._index[:]
+
+    @property
+    def ordering_(self):
+        return self.tree_.ordering_[:]
+
+    @property
+    def labels_(self):
+        return self._cluster_id[:]
+
+    def extract(self, epsilon_prime, clustering='auto', **kwargs):
+        """Performs Automatic extraction for an arbitrary epsilon.
+
+        It can also do DBSCAN equivalent and it can be run multiple times.
+
+        See extract_auto() for full description of parameters
+
+        Parameters
+        ----------
+        epsilon_prime: float or int, optional
+            Used for 'dbscan' clustering. Must be less than or equal to what
+            was used for prep and build steps
+
+        clustering: {'auto', 'dbscan'}, optional
+            Type of cluster extraction to perform; defaults to 'auto'.
+
+        Returns
+        -------
+        core_sample_indices_: array, shape (n_core_samples,)
+            The indices of the core samples.
+
+        labels_ : array, shape (n_samples,)
+            The estimated labels.
+        """
+        check_is_fitted(self, 'reachability_')
+
+        if epsilon_prime > self.eps * 5.0:
+            raise ValueError('Specify an epsilon smaller than %s. Got %s.'
+                             % (self.eps * 5, epsilon_prime))
+
+        if clustering == 'dbscan':
+            self.eps_prime = epsilon_prime
+            _extract_DBSCAN(self, epsilon_prime)
+        elif clustering == 'auto':
+            self.extract_auto(**kwargs)
+
+        self.core_sample_indices_ = \
+            np.arange(len(self.reachability_))[self._is_core == 1]
+        self.n_clusters = max(self._cluster_id)
```

Comment:
self.n_clusters_ = np.max(self._cluster_id)

but you should not have to do this outside of a fit or fit_* method.